source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs

```diff
@@ -1,79 +1,717 @@
+
+                if (!TrySetKubectl())
+                {
+                    return errorResult;
+                }
+
+                var @namespace = variables.Get(SpecialVariables.Namespace);
+                if (string.IsNullOrEmpty(@namespace))
+                {
+                    log.Verbose("No namespace provided. Using default");
+                    @namespace = "default";
+                }
+
+                if (!TrySetupContext(kubeConfig, @namespace, accountType))
+                {
+                    return errorResult;
+                }
+
+                CreateNamespace(@namespace);
+
+                var outputKubeConfig = variables.GetFlag(SpecialVariables.OutputKubeConfig);
+                if (outputKubeConfig)
+                {
+                    ExecuteKubectlCommand("config", "view");
+                }
+
+                return new CommandResult(string.Empty, 0);
+            }
+
+            bool TrySetupContext(string kubeConfig, string @namespace, string accountType)
+            {
+                var clusterUrl = variables.Get(SpecialVariables.ClusterUrl);
+                var clientCert = variables.Get("Octopus.Action.Kubernetes.ClientCertificate");
+                var eksUseInstanceRole = variables.GetFlag("Octopus.Action.AwsAccount.UseInstanceRole");
+                var podServiceAccountTokenPath = variables.Get("Octopus.Action.Kubernetes.PodServiceAccountTokenPath");
+                var serverCertPath = variables.Get("Octopus.Action.Kubernetes.CertificateAuthorityPath");
+                var isUsingPodServiceAccount = false;
+                var skipTlsVerification = variables.GetFlag(SpecialVariables.SkipTlsVerification) ? "true" : "false";
+
+                if (accountType != "AzureServicePrincipal" && string.IsNullOrEmpty(clusterUrl))
+                {
+                    log.Error("Kubernetes cluster URL is missing");
+                    return false;
+                }
+
+                string podServiceAccountToken = null;
+                string serverCert = null;
+                if (string.IsNullOrEmpty(accountType) && string.IsNullOrEmpty(clientCert) && eksUseInstanceRole == false)
+                {
+                    if (string.IsNullOrEmpty(podServiceAccountTokenPath) && string.IsNullOrEmpty(serverCertPath))
+                    {
+                        log.Error("Kubernetes account type or certificate is missing");
+                        return false;
+                    }
+
+                    if (!string.IsNullOrEmpty(podServiceAccountTokenPath))
+                    {
+                        if (File.Exists(podServiceAccountTokenPath))
+                        {
+                            podServiceAccountToken = File.ReadAllText(podServiceAccountTokenPath);
+                            if (string.IsNullOrEmpty(podServiceAccountToken))
+                            {
+                                log.Error("Pod service token file is empty");
```

Comment:
Errors with K8s logins have caused a few support tickets in the past. This would be a good opportunity to apply the [logging standards](https://github.com/OctopusDeploy/Architecture/blob/main/Steps/Logging.md) and link each of these errors to a unique code that links to the documentation.